File: typescript/src/sources.ts
import { createReadStream } from "node:fs";
import { stat } from "node:fs/promises";
import { basename } from "node:path";
import { guessMimeType } from "./mimeTypes.js";

const DEFAULT_CHUNK_SIZE = 65536;

export interface MediaSource {
  mimeType: string;
  size: number | null;
  seekable: boolean;
  chunks(): AsyncIterable<Buffer>;
}

export function fromBuffer(data: Buffer, mimeType: string): MediaSource {
  return {
    mimeType,
    size: data.length,
    seekable: true,
    async *chunks() {
      for (let i = 0; i < data.length; i += DEFAULT_CHUNK_SIZE) {
        yield data.subarray(i, i + DEFAULT_CHUNK_SIZE);
      }
    },
  };
}

export async function fromPath(path: string, mimeType?: string): Promise<MediaSource> {
  let resolvedMimeType = mimeType;
  if (!resolvedMimeType) {
    const guessed = guessMimeType(basename(path));
    if (!guessed) {
      throw new Error(
        `Could not guess mimeType for "${basename(path)}"; pass mimeType explicitly.`
      );
    }
    resolvedMimeType = guessed;
  }

  const stats = await stat(path);

  return {
    mimeType: resolvedMimeType,
    size: stats.size,
    seekable: true,
    async *chunks() {
      const stream = createReadStream(path, { highWaterMark: DEFAULT_CHUNK_SIZE });
      for await (const chunk of stream) {
        yield chunk as Buffer;
      }
    },
  };
}

export function fromAsyncIterable(
  it: AsyncIterable<Buffer>,
  mimeType: string,
  size: number | null = null
): MediaSource {
  let consumed = false;

  return {
    mimeType,
    size,
    seekable: false,
    async *chunks() {
      if (consumed) {
        throw new Error(
          "This MediaSource wraps a non-seekable stream and has already been read once."
        );
      }
      consumed = true;
      for await (const chunk of it) {
        yield chunk;
      }
    },
  };
}

File: python/src/siphon/sources.py
"""MediaSource: a uniform, chunked view over a file path, bytes, or a stream."""

from __future__ import annotations

import mimetypes
from dataclasses import dataclass
from pathlib import Path
from typing import Callable, Iterator

DEFAULT_CHUNK_SIZE = 65536


@dataclass
class MediaSource:
    mime_type: str
    size: int | None
    seekable: bool
    chunks: Callable[[], Iterator[bytes]]


def from_bytes(data: bytes, mime_type: str) -> MediaSource:
    def make_iterator() -> Iterator[bytes]:
        step = DEFAULT_CHUNK_SIZE
        for i in range(0, len(data), step):
            yield data[i : i + step]

    return MediaSource(
        mime_type=mime_type,
        size=len(data),
        seekable=True,
        chunks=make_iterator,
    )


def from_path(
    path: str | Path,
    mime_type: str | None = None,
    chunk_size: int = DEFAULT_CHUNK_SIZE,
) -> MediaSource:
    path = Path(path)
    if mime_type is None:
        guessed, _ = mimetypes.guess_type(path.name)
        if guessed is None:
            raise ValueError(
                f"Could not guess mime_type for {path.name!r}; pass mime_type explicitly."
            )
        mime_type = guessed

    size = path.stat().st_size

    def make_iterator() -> Iterator[bytes]:
        with open(path, "rb") as f:
            while chunk := f.read(chunk_size):
                yield chunk

    return MediaSource(
        mime_type=mime_type,
        size=size,
        seekable=True,
        chunks=make_iterator,
    )


def from_iterator(
    it: Iterator[bytes],
    mime_type: str,
    size: int | None = None,
) -> MediaSource:
    state = {"consumed": False}

    def make_iterator() -> Iterator[bytes]:
        if state["consumed"]:
            raise RuntimeError(
                "This MediaSource wraps a non-seekable stream and has already been read once."
            )
        state["consumed"] = True
        return it

    return MediaSource(
        mime_type=mime_type,
        size=size,
        seekable=False,
        chunks=make_iterator,
    )
